### PR TITLE
Auto-Save Feature (Solving Conflict)

### PR DESCRIPTION
**Pull request by stevenguyukai**
# Summary
## This Pull request is the same as #865 but solving the conflict with main

Closes #865

### CODE REVIEW

File: apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx:225]: Let's set the loading spinner to just the icon, not the whole component

File: apps/antalmanac/src/components/Header/SettingsMenu.tsx
Comment (KevinWu098) [inline apps/antalmanac/src/components/Header/SettingsMenu.tsx]:
```suggestion
                    <Tooltip title={<Typography>Auto Save your schedule!</Typography>}>
```

File: apps/antalmanac/src/stores/AppStore.ts
Comment (ap0nia) [inline apps/antalmanac/src/stores/AppStore.ts]: Possible strat, but not ideal:
```tsx
    autoSaveSchedule = async (userData: any) => { };
    autoSaveSchedule = async (userData: any, id: any) => { };
    autoSaveSchedule = async (userData: any, id: any, somethingelse: any) => { };
    autoSaveSchedule = async (anyArgs: any[]) => {
        type ActionType = [
            null | string,
            null | number | string | RepeatingCustomEvent | ScheduleCourse,
            null | number | string | number[],
            null | number | string | RepeatingCustomEvent
        ];
    };
```

File: apps/antalmanac/src/stores/AppStore.ts
Comment (KevinWu098) [inline apps/antalmanac/src/stores/AppStore.ts]: As we discussed in the sprint, it's best to handle this separately and be more explicit

File: apps/antalmanac/src/stores/AppStore.ts
Comment (KevinWu098) [inline apps/antalmanac/src/stores/AppStore.ts]: This is an `any`. Handle `ActionType` differently, as discussed

**Review by KevinWu098 — CHANGES_REQUESTED** [review]

File: apps/antalmanac/src/stores/AppStore.ts
Comment (KevinWu098) [inline apps/antalmanac/src/stores/AppStore.ts:380]: Can we rename this to `unsavedActions` (plural)

File: apps/antalmanac/src/stores/AppStore.ts
Comment (KevinWu098) [inline apps/antalmanac/src/stores/AppStore.ts]: Can we rename this to `unsavedActions` (plural)

**Review by KevinWu098 — CHANGES_REQUESTED** [review]

**Comment by stevenguyukai** [comment]
For only spinning the icon when saving and loading, it was not supported by the library. 

So I made a custom `SaveLoadIcon`, which will spin when `loading` is `True`. I set the default `LoadingButton` component's `loading` to `False` so the text will not spin.

> Just to document the change

**Comment by stevenguyukai** [comment]
There is a new file for the action type: `ActionTypes.ts`.

**Review by KevinWu098 — APPROVED** [review]
Great work on a fantastic feature, I think users will be really happy with this.

Code and feature looks really solid to me, so I'm going to approve it, but I'd like to discuss two (non-blocking) points at next Wednesday's meeting before merging.

1. Are all these additional DB writes a potential concern? (My guess is no, but I'll CC @EricPedley @ap0nia on this one)
2. The unsaved courses is great. However, when I load them in, and go through the `confirm` flow, it will clear: `window.localStorage.removeItem('unsavedActions');`, regardless of whether or not I ended up saving it.

**Review by KevinWu098 — APPROVED** [review]
LGTM Banger feature, merge when you get a chance 🚀🚀

File: apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx]: Is there a reason you're using a legacy class-based component instead of a functional one? Prefer functional components, and leave a comment documenting why it needs to be a class component if that's the case (i.e. maybe the button base is a class component that's incompatible with function components as its start icon?)

File: apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx]: If it's a string prop, you don't need extra braces around it.

```suggestion
            return <CircularProgress size={20} color='inherit' />;
```

File: apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx]: Could you move the last two imports from `@material-ui/core` to the destructured import statement? You didn't change any code there, but it's just weird that they're not already included. i.e.

```tsx
import {
  // ...previous imports
  Checkbox,
  FormControlLabel
} from '@material-ui/core'
```

File: apps/antalmanac/src/stores/AppStore.ts
Comment (ap0nia) [inline apps/antalmanac/src/stores/AppStore.ts:1]: My first impression of this is that you shouldn't be modifying the AppStore at all. You wrote a completely new store from scratch, so you can synchronize to that store instead of this legacy one.

File: apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx]: I think you should remove dependency on the legacy app store.

**Review by ap0nia — CHANGES_REQUESTED** [review]
Most the implementation is sensible and seems to be architected properly 👍 .
I really dislike having the dependency on the legacy appstore since there's a new store dedicated to managing state for the autosave feature. I don't have concrete implementation examples atm, but I really think it would be best to relocate all of the additional functionality from the appstore to the autosave store.

File: apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx
Comment (stevenguyukai) [reply apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx]: no, there is no specific reason. I just want to maintain consistency like how `LoadSaveButtonBase` class did that already in the file. I can change it to the functional components if that's preferred in React.

File: apps/antalmanac/src/actions/ActionTypesStore.ts
Comment (ap0nia) [inline apps/antalmanac/src/actions/ActionTypesStore.ts]:
```suggestion
```

There's no reason for the base action. The type discrimination will work without it, and the reason this type of interface would exist is to ensure classes adhere to a basic set of functionality via the `implements` (TypeScript) directive. For example

```tsx
interface BaseClass {
  doSomething: () => any
}

class MyExtendedClass implements BaseClass {
  // TS will display an error if you don't have a `doSomething` method. 
}
```

File: apps/antalmanac/src/actions/ActionTypesStore.ts
Comment (ap0nia) [inline apps/antalmanac/src/actions/ActionTypesStore.ts]: Inverse the conditional logic to utilize guard statements and reduce nesting. Also, use a type-safe function for accessing the raw string data, so that way you can easily factor in schema validation later if that's desired.

```tsx

// Somewhere else in the file
/**
  * (Somewhere else in the file)
  * This isn't type-safe within the function, 
  * but the parent scope using the function will have defined types.
  */
function parseUnsavedActionsString(unsavedActionsString: string): ActionType[] {
  try {
    return JSON.parse(unsavedActionsString)
  } catch {
    return []
  }
}

if (unsavedActionsString == null) return

if (!confirm('You have unsaved changes...')) return

const actions = parseActionsString(unsavedActionsString)

actions.forEach(action => {
  switch (action.type) {
    // ...
  }
})
```

File: apps/antalmanac/src/actions/AppStoreActions.ts
Comment (ap0nia) [inline apps/antalmanac/src/actions/AppStoreActions.ts]: This is an anti-pattern. Passing in booleans to mutate a function's behavior is like a half-baked builder pattern implementation. If you have specific functionality that you want to perform in a certain context, then take care of that in the same context instead of delegating it to another module that doesn't inherently understand it.

I'd be willing to go as far as just creating an entirely new function for auto saving, even if it means duplicating some code, just to keep these interfaces well defined. For example

```ts
export async function autosaveSchedule(userID: string, rememberMe: boolean) {
  // ...
}
```

File: apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Header/LoadSaveFunctionality.tsx]: Generally, prefer to name props literally as `props`. It's easy to understand, and there shouldn't be naming conflicts since functions are distinct entities with different namespaces. The interfaces for the props can be unique to make it clear what component the props are for.

```suggestion
function SaveLoadIcon(props: SaveLoadIconProps) {
```

File: apps/antalmanac/src/components/Header/SettingsMenu.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Header/SettingsMenu.tsx]: Use guard clauses to reduce nesting.
```suggestion
        if (!event.target.checked) return
        
        const savedUserId = window.localStorage.getItem('userID');
        
        if (!savedUserId) return

        actionTypesStore.emit('autoSaveStart');
        await saveSchedule(savedUserID, true, true);
        appStore.unsavedChanges = false;
        actionTypesStore.emit('autoSaveEnd');
```

File: apps/antalmanac/src/components/Header/SettingsMenu.tsx
Comment (ap0nia) [inline apps/antalmanac/src/components/Header/SettingsMenu.tsx]: Boxes can be styled either with the `style` prop, `sx` prop (MUIv5), or via the props directly. Generally, I prefer the former two (`style` and `sx`), since it's a consistent strategy for all components. For boxes, prefer not mixing them. i.e. You can use either, but just not at the same time.

```suggestion
                <Box style={{ display: 'flex', alignItems: 'center', gap: 4 }}>
```
Here, since gap was captured in a `style` prop, just move all the other styles to the `style` prop as well. Or vice versa if that's possible and preferrable.

File: apps/antalmanac/src/stores/AppStore.ts
Comment (ap0nia) [inline apps/antalmanac/src/stores/AppStore.ts]: This isn't enforced by the tsconfig, but prefer to indicate when things are being imported specifically for their type information, vs. as actual JS entities.

```suggestion
import type {
```

File: apps/antalmanac/src/actions/ActionTypesStore.ts
Comment (ap0nia) [inline apps/antalmanac/src/actions/ActionTypesStore.ts]: You can also use a type-safe parsing function here, like I mentioned in a different comment. Also, could you add an upper limit to the length of actions? It can be arbitrary for now, but I think it should conceptually exist. For example, a max of 1000 unsaved actions can be set.

```ts
// At the top of the module.
const MAX_UNSAVED_ACTIONS = 1_000

const unsavedActions = parseUnsavedActionsString(unsavedActionsString)
if (unsavedActions.length > MAX_UNSAVED_ACTIONS) {
  // Do something, e.g. slice of the earliest 100?
}

// If max was reached, and resolved by removing some items, 
// then you could proceed as normal
unsavedActions.push(action)
```

**Review by ap0nia — CHANGES_REQUESTED** [review]
Just some final touch ups before the final merge, keep up the good work 🚀 

Functionality seems to work.
- When autosave is off, all unsaved changes get pushed to local storage.
- When refreshing the page with autosave off and unsaved changes in local storage, I get an alert asking whether to load them.
- When autosave is on, no unsaved changes get pushed to local storage.
- When autosave is on, all actions get saved.

File: apps/antalmanac/src/actions/AppStoreActions.ts
Comment (stevenguyukai) [reply apps/antalmanac/src/actions/AppStoreActions.ts]: I added a specific function `autosaveSchedule` for auto save, and I remove the `rememberMe` in the function because autosave will not change the state of current user id in local storage.

File: apps/antalmanac/src/actions/AppStoreActions.ts
Comment (ap0nia) [inline apps/antalmanac/src/actions/AppStoreActions.ts]: You can apply guard clauses here to reduce nesting.

```suggestion
    if (userID == null) return
    userID = userID.replace(/\s+/g, '');

    if (userID.length < 0) return
    const scheduleSaveState = AppStore.schedule.getScheduleAsSaveState();
```

**Review by ap0nia — APPROVED** [review]
Looks pretty good, thanks for the work 🚀 

Feel free to merge it whenever you're ready. If you make any minor adjustments based on my latest feedback, just re-request review and I'll merge it.

**Comment by stevenguyukai** [comment]
just a quick add of the guard condition, you can just merge it if there is no problem

**Review by ap0nia — APPROVED** [review]
👍

**Pull request merged**
